### PR TITLE
[13.0][ADD] stock_picking_return_restricted_qty:New module

### PR DESCRIPTION
**Pull request by CarlosRoca13**
cc @Tecnativa TT22466

### CODE REVIEW

File: stock_picking_return_restricted_qty/wizard/stock_picking_return.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        if self.quantity > qty:
+            raise UserError(_("Return more quantities than delivered is not allowed."))
+
+    def get_returned_restricted_quantity(self, stock_move):
+        """This function is created to know how many products
+            have the person who tries to create a return picking
+            on his hand."""
+        qty = stock_move.product_qty
+        for line in stock_move.move_dest_ids.mapped("move_line_ids"):
+            if line.state in ["partially_available", "assigned"]:
```
Comment (Tardo) [inline stock_picking_return_restricted_qty/wizard/stock_picking_return.py]: Using sets is faster for this operation than lists:
```suggestion
            if line.state in {"partially_available", "assigned"}:
```

**Review by Tardo — APPROVED** [review]
:+1: Code review. Only one non-blocking comment...

File: stock_picking_return_restricted_qty/wizard/stock_picking_return.py (hunk 1 above)
Comment (CarlosRoca13) [reply stock_picking_return_restricted_qty/wizard/stock_picking_return.py]: Changes done

**Review by manuelcalerosolis — APPROVED** [review]
Code and functionality revised.

File: stock_picking_return_restricted_qty/wizard/stock_picking_return.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+class ReturnPickingLine(models.TransientModel):
+    _inherit = "stock.return.picking.line"
+
+    @api.onchange("quantity")
+    def _onchange_quantity(self):
+        qty = self.get_returned_restricted_quantity(self.move_id)
+        if self.quantity > qty:
+            raise UserError(_("Return more quantities than delivered is not allowed."))
+
+    def get_returned_restricted_quantity(self, stock_move):
```
Comment (chienandalu) [inline stock_picking_return_restricted_qty/wizard/stock_picking_return.py:46]: @api.model ?

**Review by chienandalu — APPROVED** [review]
Tested. A minor remark

File: stock_picking_return_restricted_qty/README.rst [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/stock-logistics-workflow-13-0/stock-logistics-workflow-13-0-stock_picking_return_restricted_qty
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/154/13.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module don't allows you to return more quantity of products than the
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/README.rst]:
```suggestion
This module don't allow you to return more quantity of products than the
```

File: stock_picking_return_restricted_qty/README.rst [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+delivered.
+
+.. IMPORTANT::
+   This is an alpha version, the data model and design can change at any time without warning.
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/README.rst]: This is not alpha IMO

File: stock_picking_return_restricted_qty/wizard/stock_picking_return.py (hunk 2 above)
Comment (pedrobaeza) [reply stock_picking_return_restricted_qty/wizard/stock_picking_return.py:46]: Better to not restrict it and `self` can be used any time in the future.

File: stock_picking_return_restricted_qty/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2020 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Stock Picking Return Restricted Qty",
+    "summary": "This module don't allows you to return more products"
+    "than you had bought",
+    "version": "13.0.1.0.0",
+    "development_status": "Alpha",
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/__manifest__.py]: Remove this

File: stock_picking_return_restricted_qty/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2020 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Stock Picking Return Restricted Qty",
+    "summary": "This module don't allows you to return more products"
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/__manifest__.py]:
```suggestion
    "summary": "Restrict the return to delivered quantity"
```

File: stock_picking_return_restricted_qty/readme/CONTRIBUTORS.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+* Carlos Roca <carlos.roca@tecnativa.com>
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/readme/CONTRIBUTORS.rst]: Put long format:
```suggestion
* `Tecnativa <https://www.tecnativa.com>`_:

  * Carlos Roca
```
Do you include Sergio?

**Review by Tardo — CHANGES_REQUESTED** [review]
I see the "exception" message, but the "return" action appear to continue running... is this an expected behaviour?

File: stock_picking_return_restricted_qty/readme/DESCRIPTION.rst [hunk 8]
```diff
@@ -0,0 +1,2 @@
+This module don't allows you to return more quantity of products than the
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/readme/DESCRIPTION.rst]:
```suggestion
This module don't allow you to return more quantity of products than the
```

File: stock_picking_return_restricted_qty/tests/test_stock_picking_return_restricted_qty.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
+# Copyright 2020 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo.exceptions import UserError
+from odoo.tests import common, tagged
+
+
+@tagged("post_install", "-at_install")
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/tests/test_stock_picking_return_restricted_qty.py]: Why is this needed?

File: stock_picking_return_restricted_qty/wizard/stock_picking_return.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    def _prepare_stock_return_picking_line_vals_from_move(self, stock_move):
+        val = super()._prepare_stock_return_picking_line_vals_from_move(stock_move)
+        return_lines = self.env["stock.return.picking.line"]
+        val["quantity"] = return_lines.get_returned_restricted_quantity(stock_move)
+        return val
+
+    def _create_returns(self):
+        for return_line in self.product_return_moves:
+            quantity = return_line.get_returned_restricted_quantity(return_line.move_id)
+            if return_line.quantity > quantity:
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/wizard/stock_picking_return.py]: Use `float_compare` here:

```suggestion
            if float_compare(return_line.quantity, quantity) > 0:
```

File: stock_picking_return_restricted_qty/wizard/stock_picking_return.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        """This function is created to know how many products
+            have the person who tries to create a return picking
+            on his hand."""
+        qty = stock_move.product_qty
+        for line in stock_move.move_dest_ids.mapped("move_line_ids"):
+            if line.state in {"partially_available", "assigned"}:
+                qty -= line.product_qty
+            elif line.state == "done":
+                qty -= line.qty_done
+        return qty
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/wizard/stock_picking_return.py]: You must return 0 if `qty` is negative:
```suggestion
        return max(qty, 0.0)
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by pedrobaeza** [comment]
@Tardo an exception raised aborts always the rest of the code.

**Comment by Tardo** [comment]
Ok, the exception that i get is because sets to zero the value:
error_a

**Comment by CarlosRoca13** [comment]
@pedrobaeza Changes done!

File: stock_picking_return_restricted_qty/__manifest__.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2020 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Stock Picking Return Restricted Qty",
+    "summary": "Restrict the return to delivered quantity" "than you had bought",
```
Comment (pedrobaeza) [inline stock_picking_return_restricted_qty/__manifest__.py]:
```suggestion
    "summary": "Restrict the return to delivered quantity"
```

File: stock_picking_return_restricted_qty/wizard/stock_picking_return.py (hunk 10 above)
Comment (pedrobaeza) [reply stock_picking_return_restricted_qty/wizard/stock_picking_return.py]: This is still pending

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by CarlosRoca13** [comment]
@pedrobaeza done!

File: stock_picking_return_restricted_qty/__manifest__.py (hunk 12 above)
Comment (pedrobaeza) [reply stock_picking_return_restricted_qty/__manifest__.py]: This is still incorrect. You are not buying anything, but returning what has been delivered.

Comment (CarlosRoca13) [reply stock_picking_return_restricted_qty/__manifest__.py]: Ok true

Comment (CarlosRoca13) [reply stock_picking_return_restricted_qty/__manifest__.py]: Solve it yet

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
/ocabot merge nobump

Please check next time that you have attended all comments, not one by one.
